File: typescript/leetcode/0015-3sum/solution01.ts
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
export function threeSum(nums: number[]): number[][] {
  const sum: number[][] = []

  // O(n**3)
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      for (let k = j + 1; k < nums.length; k++) {
        if (nums[i] + nums[j] + nums[k] !== 0) {
          continue
        }

        let duplicated = false
        const possible = [nums[i], nums[j], nums[k]]
        for (let l = 0; l < sum.length; l++) {
          const triplets = sum[l]
          if (same(triplets, possible)) {
            duplicated = true
            break
          }
        }
        if (!duplicated) {
          sum.push(possible)
        }
      }
    }
  }

  return sum
}

function same(a: number[], b: number[]): boolean {
  a.sort((a, b) => a - b)
  b.sort((a, b) => a - b)
  return a[0] === b[0] && a[1] === b[1] && a[2] === b[2]
}

File: python/leetcode/0015-3sum/solution01.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue  # to avoid duplication

            lo = i + 1
            hi = len(nums) - 1

            complement = 0 - nums[i]
            while lo < hi:
                if (nums[lo] + nums[hi]) == complement:
                    result.append([nums[i], nums[lo], nums[hi]])

                    # to avoid duplication
                    while nums[lo] == nums[lo + 1]:
                        lo += 1
                    while nums[hi] == nums[hi - 1]:
                        hi -= 1

                    lo += 1
                    hi -= 1
                elif nums[lo] + nums[hi] < complement:
                    lo += 1
                else:
                    hi -= 1

        return result
